BidCat/API/Auction.cs

```diff
@@ -9,40 +9,179 @@ namespace BidCat.API
+			bool allowVisibleLowering = true)
+		{
+			if (amount < 1)
+			{
+				throw new ApiError("Amount must be a number above 0");
+			}
+
+			int? previousBid = null;
+			if (lotDict.TryGetValue(item, out _))
+			{
+				previousBid = lotDict[item].FirstOrDefault(x => x.Item1 == user)?.Item2;
+			}
+
+			bool alreadyBid = previousBid != null;
+
+			if (!replace && alreadyBid)
+				throw new AlreadyBidError("There is already a bid from that user on that item.");
+			if (replace && !alreadyBid)
+				throw new NoExistingBidError("There is no bid from that user on that item which could be replaced.");
+			if (replace && previousBid == amount)
+				return;
+			int neededMoney = amount;
+			if (replace)
+				neededMoney -= (int)previousBid;
+			int availableMoney = await bank.GetAvailableMoney(user);
+			if (neededMoney > availableMoney)
+				throw new InsufficientMoneyError($"Can't afford to bid {neededMoney}, only {availableMoney} available.");
+			if (replace && amount < previousBid && !allowVisibleLowering)
+			{
+				Dictionary<string, object> winner = await GetWinner();
+				if ((string)winner["item"] != item)
+					throw new VisiblyLoweredError();
+				int headroom = (int) winner["total_bid"] - (int) winner["total_charge"];
+				int decrease = (int)previousBid - amount;
+				if (decrease > headroom)
+					throw new VisiblyLoweredError();
+			}
+
+			UpdateLastChange(item);
+			if (!lotDict.ContainsKey(item))
+				lotDict[item] = new List<Tuple<int, int>>();
+			lotDict[item].Add(new Tuple<int, int>(user, amount));
+		}
+
+#pragma warning disable 1998
+		public async Task<List<Tuple<string, List<Tuple<int, int>>>>> GetAllBidsOrdered()
+		{
+			List<Tuple<string, List<Tuple<int, int>>>> result = new List<Tuple<string, List<Tuple<int, int>>>>();
+			foreach (KeyValuePair<string, List<Tuple<int, int>>> kvp in lotDict)
+			{
+				Tuple<string, List<Tuple<int, int>>> tuple =
+					new Tuple<string, List<Tuple<int, int>>>(kvp.Key, kvp.Value);
+				result.Add(tuple);
+			}
+
+			return result.OrderByDescending(x => x.Item2.Sum(y => y.Item2)).ToList();
+		}
+#pragma warning restore 1998
+
+		public async Task<Dictionary<string, object>> GetWinner(bool discountLatter = false)
+		{
+			List<Tuple<string, List<Tuple<int, int>>>> bids = await GetAllBidsOrdered();
+			if (!bids.Any())
+				return null;
+			(string winningItem, List<Tuple<int, int>> winningBids) = bids[0];
+			int secondBid = 0;
+			if (bids.Count > 1)
+			{
+				List<Tuple<int, int>> secondItemBids = bids[1].Item2;
+				secondBid = secondItemBids.Sum(x => x.Item2);
+			}
+
+			int totalBid = winningBids.Sum(x => x.Item2);
+			int overpaid = Math.Max(0, totalBid - secondBid - 1);
+			int totalCharge = totalBid - overpaid;
+			
+			List<Tuple<int, int>> moneyOwed = (from userAmountTuple in winningBids.OrderByDescending(x => x.Item2) let percentage = userAmountTuple.Item2 / (double) totalBid select new Tuple<int, int>(userAmountTuple.Item1, (int) Math.Ceiling(totalCharge * percentage))).ToList();
```

Comment:
You use this structure a lot to represent bids. Time to make it a class or struct instead of an anonymous tuple.